scripts/src/java/org/oppia/android/scripts/coverage/CoverageReporter.kt

```diff
@@ -31,200 +38,535 @@ class CoverageReporter(
-            <td class="$lineClass">$line</td>
-        </tr>
-      """.trimIndent()
     }
 
-    htmlContent += """
-        </tbody>
-      </table>
-    </body>
-    </html>
-    """.trimIndent()
-
-    return Pair(computedCoverageRatio, htmlContent)
+    if (failureReports.isNotEmpty()) {
+      println(
+        """
+        |
+        |COVERAGE FAILURE REPORT:
+        |-----------------------
+        """.trimMargin().prependIndent("  ")
+      )
+      println(failureReports)
+    }
   }
 
-  private fun computeCoverageRatio(): Float {
-    return if (coverageReport.linesFound != 0) {
-      coverageReport.linesHit.toFloat() / coverageReport.linesFound.toFloat()
-    } else {
-      0f
-    }
+  private fun generateTableRows(
+    reports: List<CoverageReport>,
+    statusSymbol: String
+  ): String {
+    return reports
+      .mapNotNull { report ->
+        val details = report.details
+        val filePath = details.filePath
+        val totalLinesFound = details.linesFound
+        val totalLinesHit = details.linesHit
+        val exemptionPercentage = testFileExemptionList[filePath]
+          ?.overrideMinCoveragePercentRequired
+          ?: MIN_THRESHOLD
+        val coveragePercentage = calculateCoveragePercentage(
+          totalLinesHit, totalLinesFound
+        )
+        val formattedCoveragePercentage = "%2.2f".format(coveragePercentage)
+
+        "| ${getFilenameAsLink(filePath)} | $formattedCoveragePercentage% | " +
+          "$totalLinesHit / $totalLinesFound | $statusSymbol | $exemptionPercentage% |"
+      }
+      .joinToString(separator = "\n")
   }
 }
 
+/** Corresponds to status of the coverage analysis. */
+enum class CoverageCheck {
+  /** Indicates successful generation of coverage retrieval for a specified file. */
+  PASS,
+  /** Indicates failure or anomaly during coverage retrieval for a specified file. */
+  FAIL
+}
+
 /** Represents the different types of formats available to generate code coverage reports. */
 enum class ReportFormat {
   /** Indicates that the report should be formatted in .md format. */
   MARKDOWN,
   /** Indicates that the report should be formatted in .html format. */
   HTML
 }
+
+private fun calculateCoveragePercentage(linesHit: Int, linesFound: Int): Float {
+  return linesFound.takeIf { it > 0 }
+    ?.let { (linesHit.toFloat() / it * 100).toFloat() }
+    ?: 0f
+}
+
+private fun getReportOutputPath(
+  repoRoot: String,
+  filePath: String,
+  reportFormat: ReportFormat
+): String {
+  val fileWithoutExtension = filePath.substringBeforeLast(".")
+  val defaultFilename = when (reportFormat) {
+    ReportFormat.HTML -> "coverage.html"
+    ReportFormat.MARKDOWN -> "coverage.md"
+  }
+  return "$repoRoot/coverage_reports/$fileWithoutExtension/$defaultFilename"
+}
+
+private fun getFilenameAsLink(filePath: String): String {
+  val oppiaDevelopGitHubLink = "https://github.com/oppia/oppia-android/tree/develop"
```

Comment:
This is a nice idea, but it's also a bit tricky because the actual file under test will be specific to the PR being run or the current branch. I'm torn here--we either include a link that is most often wrong, or we exclude it entirely. What do you think?

Reply:
Ahh I did not actually realize that these file links should correspond to their respective PRs diff files. 

Uh... we can either try to find the feature branch name and use the same source path to actually locate the pr file - https://github.com/oppia/oppia-android/blob/code_coverage_list_of_files/scripts/src/java/org/oppia/android/scripts/coverage/CoverageReporter.kt

```
[base link]-[feature-branch]-[source-path]
```
But I guess this will only work if the branch is in upstream (which makes this possibility obsolete) 

---

Don't know if these links would be really useful to have it as a mandatory thing but I liked it as links and still do (as it was better to rather have it as short filename links than as a long text of its path), will look if there are any possibility of at least grabbing the pr's diff files. 

Will update if I can find anything useful.

Reply:
Per discussion I think we want to remove these now, right? You're correct that it's really not simple since local vs. develop vs. a PR will all have different link contexts, and it's quite difficult to get it correct in each circumstance. :)

Reply:
Yeah. this is the one thing that is left to do in this PR (apart from stuff with the recent review), will modify the template and update here (and PR description) with the most recent version of the report template.

Reply:
Updated the md report replacing the links with details-summary tags:

**Template:**

## Coverage Report

### Results
Number of files assessed: 5
Overall Coverage: **94.26%**
Coverage Analysis: **FAIL** :x:
##

### Failure Cases

| File | Failure Reason |
|------|----------------|
| File.kt | No appropriate test file found for File.kt. |

### Failing coverage

| File | Coverage | Lines Hit | Status | Min Required |
|------|:--------:|----------:|:------:|:------------:|
| <details><summary>RunCoverage.kt</summary>scripts/src/java/org/oppia/android/scripts/coverage/RunCoverage.kt</details> | 51.38% | 148 / 288 | :x: | 70% |
| <details><summary>MathModel.kt</summary>utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt</details> | 77.78% | 7 / 9 | :x: | 80% _*_ |
| <details><summary>MultipleChoiceInputModule.kt</summary>domain/src/main/java/org/oppia/android/domain/classify/rules/multiplechoiceinput/MultipleChoiceInputModule.kt</details> | 10.00% | 1 / 10 | :x: | 30% _*_ |

>**_*_** represents tests with custom overridden pass/fail coverage thresholds

### Passing coverage

<details>
<summary>Files with passing code coverage</summary><br>

| File | Coverage | Lines Hit | Status | Min Required |
|------|:--------:|----------:|:------:|:------------:|
| <details><summary>MathTokenizer.kt</summary>utility/src/main/java/org/oppia/android/util/math/MathTokenizer.kt</details> | 94.26% | 197 / 209 | :white_check_mark: | 70% |
</details>

### Exempted coverage
<details><summary>Files exempted from coverage</summary> <br>
<details><summary>ActivityComponent.kt</summary>app/src/main/java/org/oppia/android/app/activity/ActivityComponent.kt</details>
<details><summary>TestExemptedFile.kt</summary>app/src/main/java/org/oppia/android/app/activity/TestExemptedFile.kt</details>
<details><summary>SourceIncompatible.kt</summary>app/src/main/java/org/oppia/android/app/activity/SourceIncompatible.kt</details>

Reply:
This looks so nice! Thanks for making the changes.